query-parser/src/parse.rs

```diff
@@ -83,320 +96,348 @@ error_chain! {
-            v.unwrap_or(None)
-        } else {
-            None
-        }
-    }).parse_stream(input)
-});
-
-def_value_parser_fn!(Query, arguments, Vec<FnArg>, input, {
-    (many::<Vec<FnArg>, _>(Query::fn_arg()))
-    .skip(eof())
-    .parse_stream(input)
+            (optional(Query::source_var()),              // src
+             Where::pattern_non_value_place(),           // e
+             Where::pattern_non_value_place(),           // a
+             optional(Where::pattern_value_place()),     // v
+             optional(Where::pattern_non_value_place())) // tx
+                .and_then(|(src, e, a, v, tx)| {
+                    let v = v.unwrap_or(PatternValuePlace::Placeholder);
+                    let tx = tx.unwrap_or(PatternNonValuePlace::Placeholder);
+
+                    // Pattern::new takes care of reversal of reversed
+                    // attributes: [?x :foo/_bar ?y] turns into
+                    // [?y :foo/bar ?x].
+                    //
+                    // This is a bit messy: the inner conversion to a Pattern can
+                    // fail if the input is something like
+                    //
+                    // ```edn
+                    // [?x :foo/_reversed 23.4]
+                    // ```
+                    //
+                    // because
+                    //
+                    // ```edn
+                    // [23.4 :foo/reversed ?x]
+                    // ```
+                    //
+                    // is nonsense. That leaves us with a nested optional, which we unwrap here.
+                    Pattern::new(src, e, a, v, tx)
+                        .map(WhereClause::Pattern)
+                        .ok_or(combine::primitives::Error::Expected("pattern".into()))
+                }))
 });
 
-def_value_parser_fn!(Where, clause, WhereClause, input, {
-    choice([Where::pattern(),
-            Where::pred(),
+def_parser!(Where, clause, WhereClause, {
+    choice([try(Where::pattern()),
             // It's either
             //   (or-join [vars] clauses…)
             // or
             //   (or clauses…)
             // We don't yet handle source vars.
-            Where::or_join_clause(),
-            Where::or_clause(),
-    ]).parse_stream(input)
+            try(Where::or_join_clause()),
+            try(Where::or_clause()),
+
+            try(Where::pred()),
+    ])
 });
 
-def_value_parser_fn!(Where, clauses, Vec<WhereClause>, input, {
+def_parser!(Where, clauses, Vec<WhereClause>, {
     // Right now we only support patterns and predicates. See #239 for more.
     (many1::<Vec<WhereClause>, _>(Where::clause()))
-    .skip(eof())
-    .parse_stream(input)
 });
 
-pub struct Find<I>(::std::marker::PhantomData<fn(I) -> I>);
+pub struct Find;
 
-def_value_parser_fn!(Find, period, (), input, {
-    matches_plain_symbol!(".", input)
+/// TODO: extract macro for matching these `PlainSymbol` instances.
```

Comment:
Done.  Just FYI, `$struct` and `$match` are both forbidden, since they parse as keywords.